### ISSUE TITLE
Duplicate entry race conditions

### ISSUE THREAD

**Issue by Christiaanben** [tinode/chat#818]
### Creating two entries in quick succession creates a duplicate entry for the `messages.messages_topic_seqid` constraint
We are using tinode with an external bot that responds almost immediately. When these messages are sent to tinode in quick succession they get assigned duplicate seq_id. For some reason it doesn't just check what the last seq_id was and keeps trying to insert new messages with the same seq_id.

#### Server-side
- [x] Your own setup:
  * platform: Linux (docker)
  * version of Tinode server: `0.18.4`
  * database backend: MySQL
  * standalone

#### Client-side
- [x] Your own client. Describe it:
  * Transport: gRPC
  * Programming language: python

### Steps to reproduce
Create many messages in quick succession (asynchronously) until you encounter `Error 1062: Duplicate entry`. Future messages get stuck with this same error

### Expected behaviour
Messages should be inserted with a new seq_id

### Actual behaviour
It keeps trying to save using the same seq_id

### Server-side log
```
I2023/01/05 12:30:12 in: '
{
    "sub": {
        "id": "73340",
        "topic": "grpwBvslNZ8xpU",
        "get": {
            "data": {
                "since": 584,
                "limit": 24
            },
            "desc": {
                "ims": "2023-01-04T13:45:55.632Z"
            },
            "del": {
                "since": 1
            },
            "what": "data sub desc del"
        }
    }
}
' sid='ddrQ9eZjFaI' uid='8FU4Wa59ioA'
I2023/01/05 12:31:14 in: '
{
    "sub": {
        "get": {
            "data": {
                "limit": 24,
                "since": 584
            },
            "desc": {
                "ims": "2022-12-09T12:46:42.988Z"
            },
            "sub": {
                "ims": "2022-12-09T12:46:43.400Z"
            },
            "what": "desc sub data del"
        },
        "id": "88606",
        "topic": "grpwBvslNZ8xpU"
    }
}
' sid='qUzv9HfL2bk' uid='8FU4Wa59ioA'
I2023/01/05 12:31:14 in: '
{
    "sub": {
        "get": {
            "data": {
                "limit": 24,
                "since": 584
            },
            "desc": {
                "ims": "2022-12-09T12:46:42.988Z"
            },
            "sub": {
                "ims": "2022-12-09T12:46:43.400Z"
            },
            "what": "desc sub data del"
        },
        "id": "88607",
        "topic": "grpwBvslNZ8xpU"
    }
}
' sid='qUzv9HfL2bk' uid='8FU4Wa59ioA'
I2023/01/05 12:31:14 in: '
{
    "pub": {
        "content": "Trigger Starter",
        "id": "88608",
        "noecho": true,
        "topic": "grpwBvslNZ8xpU"
    }
}
' sid='qUzv9HfL2bk' uid='8FU4Wa59ioA'
W2023/01/05 12:31:14 topic[grpwBvslNZ8xpU]: failed to save message: Error 1062: Duplicate entry 'grpwBvslNZ8xpU-579' for key 'messages.messages_topic_seqid'
I2023/01/05 12:31:17 in: '
{
    "pub": {
        "content": "Hello",
        "id": "88609",
        "noecho": true,
        "topic": "grpwBvslNZ8xpU"
    }
}
' sid='qUzv9HfL2bk' uid='8FU4Wa59ioA'
W2023/01/05 12:31:17 topic[grpwBvslNZ8xpU]: failed to save message: Error 1062: Duplicate entry 'grpwBvslNZ8xpU-579' for key 'messages.messages_topic_seqid'
I2023/01/05 12:31:39 grpc in: sub:{id:"LS3T63T4gf7zAtFiYkRwQJ" topic:"grpwBvslNZ8xpU"} hOfeHlO-1eE
I2023/01/05 12:31:39 grpc in: pub:{topic:"grpwBvslNZ8xpU" no_echo:true head:{key:"mime" value:"\"text/x-drafty\""} head:{key:"origin" value:"\"messenger\""} content:"{\"txt\": \"Hello\", \"fmt\": [], \"ent\": []}"} extra:{on_behalf_of:"usr4fjcVLUIaxg"} hOfeHlO-1eE
W2023/01/05 12:31:39 topic[grpwBvslNZ8xpU]: failed to save message: Error 1062: Duplicate entry 'grpwBvslNZ8xpU-579' for key 'messages.messages_topic_seqid'
I2023/01/05 12:31:44 grpc in: sub:{id:"WNKFTjkfqfJ34QrFrmybQP" topic:"grpwBvslNZ8xpU"} itrGCC_1nS4
I2023/01/05 12:31:44 grpc in: pub:{topic:"grpwBvslNZ8xpU" no_echo:true head:{key:"mime" value:"\"text/x-drafty\""} head:{key:"origin" value:"\"messenger\""} content:"{\"txt\": \"Anyone there?\", \"fmt\": [], \"ent\": []}"} extra:{on_behalf_of:"usr4fjcVLUIaxg"} itrGCC_1nS4
W2023/01/05 12:31:44 topic[grpwBvslNZ8xpU]: failed to save message: Error 1062: Duplicate entry 'grpwBvslNZ8xpU-579' for key 'messages.messages_topic_seqid'
I2023/01/05 12:31:48 grpc in: sub:{id:"KLzMTQaiBv2xbqxzCHYHk9" topic:"grpwBvslNZ8xpU"} BJ-iHD38rTY
I2023/01/05 12:31:48 grpc in: pub:{topic:"grpwBvslNZ8xpU" no_echo:true head:{key:"mime" value:"\"text/x-drafty\""} head:{key:"origin" value:"\"messenger\""} content:"{\"txt\": \"Testing my messages\", \"fmt\": [], \"ent\": []}"} extra:{on_behalf_of:"usr4fjcVLUIaxg"} BJ-iHD38rTY
W2023/01/05 12:31:48 topic[grpwBvslNZ8xpU]: failed to save message: Error 1062: Duplicate entry 'grpwBvslNZ8xpU-579' for key 'messages.messages_topic_seqid'
```

**Comment by or-else** [comment]
1. You are sending two subscribe requests to the same topic from the same session:
```
"id": "88606", "topic": "grpwBvslNZ8xpU" sid='qUzv9HfL2bk' uid='8FU4Wa59ioA'
"id": "88607", "topic": "grpwBvslNZ8xpU" sid='qUzv9HfL2bk' uid='8FU4Wa59ioA'
```
Can you check if this is what's triggering the data race?

2. Can you produce a small python script which triggers the problem so I can run it on my end?

3. It seems like you are running a pretty old version of the server. There were quite a few changes since then:
https://github.com/tinode/chat/compare/v0.18.4...v0.20.4
Can you verify that the problem is still present in 20.4+? It probably is, but it would be nice to know. Also, we are unlikely to patch the 18 branch. The fixes will go to 20 or 21.

**Comment by or-else** [comment]
I'm pretty sure the problem is due to double subscribe. Here is a quick fix:
https://github.com/tinode/chat/commit/9739b10d0d98e36cffad6986353d53b858c953b8
Can you verify that it fixes the problem?

**Comment by or-else** [comment]
The fix was released in 0.21.0-rc1
